### PR TITLE
mgr/orchestrator: make use of @CLICommand

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

removes about 100 LOC from the module.

- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug
- [ ] Now depends on #26171

### CODE REVIEW

File: src/pybind/mgr/orchestrator_cli/module.py [hunk 1]
```diff
@@ -127,8 +33,12 @@ def _select_orchestrator(self):
 
         return o
 
-
-    def _list_devices(self, cmd):
+    @CLIReadCommand('orchestrator device ls',
+                    "name=host,type=CephString,req=false "
```
Comment (epuertat) [inline src/pybind/mgr/orchestrator_cli/module.py]:
```suggestion
                    "name=host,type=CephString,n=N,req=false "
```
With that you can get a list of hosts for free.

File: src/pybind/mgr/orchestrator_cli/module.py [hunk 2]
```diff
@@ -179,14 +83,18 @@ def _list_devices(self, cmd):
 
             return HandleCommandResult(stdout=result)
 
-    def _list_services(self, cmd):
-        hostname = cmd.get('host', None)
-        svc_id = cmd.get('svc_id', None)
-        svc_type = cmd.get('svc_type', None)
+    @CLIReadCommand('orchestrator service ls',
+                    "name=host,type=CephString,req=false "
+                    "name=svc_type,type=CephString,req=false "
```
Comment (epuertat) [inline src/pybind/mgr/orchestrator_cli/module.py]: Is `svc_type` not a good candidate for a `CephChoices`?

File: src/pybind/mgr/orchestrator_cli/module.py [hunk 3]
```diff
@@ -1,124 +1,30 @@
...
+        super(NoOrchestrator, self).__init__("No orchestrator configured (try "
+                                             "`ceph orchestrator set backend`)")
+
+
+def handle_exceptions(func):
+    def inner(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except (NoOrchestrator, ImportError) as e:
+            return HandleCommandResult(-errno.ENOENT, stderr=str(e))
```
[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator_cli/module.py]: This is meant to explicitly catch `NoOrchestrator`. But yea, having something similar in `mgr_module` probably would make sense.

File: src/pybind/mgr/orchestrator_cli/module.py [hunk 4]
```diff
@@ -127,8 +33,12 @@ def _select_orchestrator(self):
...
         return o
 
-
-    def _list_devices(self, cmd):
+    @CLIReadCommand('orchestrator device ls',
+                    "name=host,type=CephString,req=false "
+                    "name=format,type=CephChoices,strings=json|plain,req=false",
+                    'List devices on a node')
+    @handle_exceptions
+    def _list_devices(self, host=None, format='plain'):
```
[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator_cli/module.py]: Yeah, I also noticed that `@CLI*Command` expects the first argument to be of type `MgrModule`:

https://github.com/ceph/ceph/blob/1b206291f839442f564edccf00dcee07b4974c9b/src/pybind/mgr/mgr_module.py#L830

AFAICS, `@CLI*Command` was mean to be compatible to be used with module methods.

File: src/pybind/mgr/orchestrator_cli/module.py [hunk 5]
```diff
@@ -1,124 +1,30 @@
...
 
 class NoOrchestrator(Exception):
-    pass
+    def __init__(self):
+        super(NoOrchestrator, self).__init__("No orchestrator configured (try "
+                                             "`ceph orchestrator set backend`)")
+
+
+def handle_exceptions(func):
+    def inner(*args, **kwargs):
```
[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator_cli/module.py]: fixed

File: src/pybind/mgr/orchestrator_cli/module.py (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator_cli/module.py]: added

File: src/pybind/mgr/orchestrator_cli/module.py (hunk 2 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator_cli/module.py]: yep.

**Comment by tchaikov** [comment]
@sebastian-philipp needs rebase

File: src/pybind/mgr/orchestrator.py [hunk 6]
```diff
@@ -694,26 +694,57 @@ class InventoryDevice(object):
...
                     available=self.available, dev_id=self.dev_id,
-                    **self.extended)
+                    extended=self.extended)
+
+    @classmethod
+    def from_ceph_volume_inventory(cls, data):
+        dev = InventoryDevice()
+        dev.id = data["path"]
+        dev.type = 'hdd' if data["sys_api"]["rotational"] == "1" else 'sdd/nvme'
+        dev.size = data["sys_api"]["size"]
```
Comment (jmolmo) [inline src/pybind/mgr/orchestrator.py]: why not
dev.size = ["sys_api"]["human_readable_size"] 
?

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: We have to stick to the Interface. If rook, ssh and deepsea are adding an int in bytes, ansible cannot use that field as `human_readable_size`

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: InventoryDevice predates c-v inventory. as described in InventoryDevice is supposed to be of type `int`. rook and deepsea already populate `size` as such.

Comment (jmolmo) [reply src/pybind/mgr/orchestrator.py]: ok. Probably a human friendly output will be requested at some time... it can be produced easily.
LGTM

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: Ive just added format_bytes: 

https://github.com/ceph/ceph/pull/26094/commits/81dc75be10a6cf648932b876b99f843284942f26#diff-dfbc76d00f82320b0e6ac732319aa545R768

Comment (jmolmo) [reply src/pybind/mgr/orchestrator.py]: Thx!

**Pull request merged**